sw/host/ot_certs/src/offsets.rs

```diff
@@ -0,0 +1,199 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{bail, Context, Result};
+use std::collections::HashMap;
+
+use crate::template::{Conversion, Value, VariableType};
+use crate::types::{convert_type, Type};
+
+#[derive(Debug, Clone)]
+pub struct CertificateOffset {
+    /// Offset into the DER.
+    pub offset: usize,
+    /// Type of the variable in the DER.
+    pub target_type: VariableType,
+    /// Conversion that needs to be applied to the original variable.
+    pub conversion: Option<Conversion>,
+}
+
+#[derive(Debug, Clone)]
+pub struct CertificateVariable {
+    pub name: String,
+    pub source_type: VariableType,
+    pub offsets: Vec<CertificateOffset>,
+}
+
+/// X509 certificate in DER form where the offset to variables has been recorded.
+#[derive(Debug, Clone)]
+pub struct CertificateWithVariables {
+    /// Name of the certificate, will be used to generate names in the template.
+    pub name: String,
+    /// Certificate in binary form.
+    pub cert: Vec<u8>,
+    /// List of offsets where the variables are used. A variable can appear multiple
+    /// times with different types.
+    pub variables: Vec<CertificateVariable>,
+}
+
+struct GeneratorVariable {
+    hidden: bool,
+    source_type: VariableType,
+    values: Vec<Vec<u8>>,
+    offsets: Vec<CertificateOffset>,
+}
+
+impl GeneratorVariable {
+    fn new(source_type: VariableType) -> GeneratorVariable {
+        GeneratorVariable {
+            hidden: false,
+            source_type,
+            values: Vec::new(),
+            offsets: Vec::new(),
+        }
+    }
+}
+
+// The generators holds the information about the offset to the variables
+// in the certificate and the randomly generated values that we used during generation.
+pub struct OffsetGenerator {
+    variables: HashMap<String, GeneratorVariable>,
+}
+
+// Super inefficient function to search for needle in a haystack.
+fn find_subsequence(haystack: &[u8], needle: &[u8]) -> Option<usize> {
+    haystack
+        .windows(needle.len())
+        .position(|window| window == needle)
+}
+
+impl OffsetGenerator {
+    pub fn new(variables: &HashMap<String, VariableType>) -> OffsetGenerator {
+        let variables = variables
+            .iter()
+            .map(|(name, source_type)| (name.clone(), GeneratorVariable::new(*source_type)))
+            .collect();
+        OffsetGenerator { variables }
+    }
+
+    pub fn add_variable<T>(&mut self, bytes: Vec<u8>, target_type: VariableType, val: &Value<T>) {
+        match val {
+            Value::Literal(_) => {
+                panic!("add_variable can only be called on variables, not literals")
+            }
+            Value::Variable { var: name, convert } => {
```

Comment:
```suggestion
        let Value::Variable { var: name, convert } = val else {
            panic!("add_variable can only be called on variables, not literals");
        };
```

Also, maybe an error would be better than a panic here